SharedSource/Core/Utilities/ArgumentUtility.cs

```diff
@@ -19,7 +19,7 @@
 using System.Diagnostics;
 using System.Reflection;
 using JetBrains.Annotations;
-
+#nullable disable

```

Comment:
At the current state, `ArgumentUtility` does not provide the needed flow enhancements.

```csharp
void Do (object a) {}

void A (object? a)
{
  ArgumentUtility.CheckNotNull ("a", a);

  Do (a); // [CS8604] Possible null reference argument for parameter 'a' in 'void A(object a)'.
}
```

I guess our main use case is providing runtime argument checks for **non-nullable** parameters, but the flow should still be enhanced.

Reply:
Unforturnately I can't comment on unchanged lines, so I'll post everything here.

```csharp
[AssertionMethod]
[return: NotNullIfNotNull ("actualValue")]
public static T CheckNotNull<T> (
    [InvokerParameterName] string argumentName,
    [System.Diagnostics.CodeAnalysis.NotNull, AssertionCondition (AssertionConditionType.IS_NOT_NULL), NoEnumeration] T actualValue)
{
  // ReSharper disable CompareNonConstrainedGenericWithNull
  if (actualValue == null)
      // ReSharper restore CompareNonConstrainedGenericWithNull
    throw new ArgumentNullException (argumentName);

  return actualValue;
}
```

`[return: NotNullIfNotNull ("actualValue")]` is needed on every method doing null checks and returning the value if not null.

`[System.Diagnostics.CodeAnalysis.NotNull]` fixes the wrong flow addressed in the first comment. This needs to be added on every parameter that is null checked.